near/src/runtime.rs

```diff
@@ -92,63 +100,21 @@ impl NightshadeRuntime {
-        let (stake_change, validator_to_index, total_gas_used) = {
-            let epoch_info = epoch_manager.get_epoch_info(&epoch_id)?;
-            (&epoch_info.stake_change, &epoch_info.validator_to_index, epoch_info.total_gas_used)
-        };
-        let prev_keys: BTreeSet<_> = prev_stake_change.keys().collect();
-        let keys: BTreeSet<_> = stake_change.keys().collect();
-
-        // calculate reward for validators
-        // TODO: include storage rent
-        let total_tx_fee = gas_price * total_gas_used as u128;
-        println!("total_tx_fee: {}", total_tx_fee);
-        // first order taylor approximation of the inflation
-        let max_inflation = self.genesis_config.max_inflation_rate as u128
-            * total_supply
-            * self.genesis_config.epoch_length as u128
-            * MIN_BLOCK_PRODUCTION_DELAY as u128
-            / (100 * 365 * 24 * 60 * 60);
-        println!("max inflation: {}", max_inflation);
-        let total_reward = max(
-            max_inflation,
-            (100 - self.genesis_config.developer_reward_percentage) as u128 * total_tx_fee / 100,
-        );
-        // TODO: add to protocol treasury
-        let protocol_reward =
-            self.genesis_config.protocol_reward_percentage as u128 * total_reward / 100;
-        let validator_total_reward = total_reward - protocol_reward;
-        let num_validators = validator_to_index.len() - slashed.len();
-        let reward = validator_total_reward / num_validators as u128;
-        println!("reward: {}", reward);
-        println!(
-            "prev_prev_stake_change {:?} prev_stake_change: {:?} stake_change {:?}",
-            prev_prev_stake_change, prev_stake_change, stake_change
-        );
+        let (stake_info, validator_reward) = epoch_manager.compute_stake_return_info(block_hash)?;
 
-        for account_id in prev_keys.union(&keys) {
-            let account: Option<Account> = get_account(&state_update, account_id);
+        for (account_id, max_of_stakes) in stake_info {
+            let account: Option<Account> = get_account(state_update, &account_id);
             if let Some(mut account) = account {
-                let new_stake = *stake_change.get(*account_id).unwrap_or(&0);
-                let prev_stake = *prev_stake_change.get(*account_id).unwrap_or(&0);
-                let prev_prev_stake = *prev_prev_stake_change.get(*account_id).unwrap_or(&0);
-                let max_of_stakes =
-                    vec![prev_prev_stake, prev_stake, new_stake].into_iter().max().unwrap();
+                if let Some(reward) = validator_reward.get(&account_id) {
+                    println!(
+                        "account {} adding reward {} to stake {}",
+                        account_id, reward, account.stake
+                    );
+                    account.stake += *reward;
+                }
+
                 if account.stake < max_of_stakes {
```

Comment:
let's make this an assert, because it will crash on next line anyway, but at least for assert you can format and add more debug info